08-compiler-features.tex

```diff
@@ -36,6 +139,143 @@ \section{Dead Code Elimination}
+
+In our case, that would be String. Except that in the cases of an anonymous object or a class instance, the compiler will instead list the different fields in an XML and print them to the standard error output. Here's an example of output:
+
+\begin{lstlisting}
+<list>
+<i n="length"><t>Int</t><d>...</d></i>
+<i n="charAt"><t>index : Int -> String</t><d>...</d></i>
+<i n="charCodeAt"><t>index : Int -> Int</t><d>...</d></i>
+<i n="indexOf">
+  <t>value : String -> ?startIndex : Int -> Int</t>
+  <d>...</d>
+</i>
+...
+</list>
+\end{lstlisting}
+
+In that case, all public methods and fields of String are listed. This information can directly be used by the IDE to provide code hints and completion.
+
+\paragraph{Functions completion}
+This completion mechanism works with both dot and opening parenthesis, so you can get type informations about fields list and function call arguments types:
+
+\begin{lstlisting}
+class Test {
+    public static function main() {
+        trace("Hello".split(|
+    }
+}
+\end{lstlisting}
+
+This will give you the following result: \expr{<type>delimiter : String -> Array<String></type>} \_(Please note that in real output, the content of the type not is html-escaped)\_
+
+\paragraph{Package completion}
+It also works with packages, by scanning the .hx files available in the classpath:
+
+\begin{lstlisting}
+import haxe.|
+Will return :
+
+<list>
+    <i n="BaseCode"><t></t><d></d></i>
+    <i n="EnumFlags"><t></t><d></d></i>
+    ...
+    <i n="io"><t></t><d></d></i>
+    <i n="macro"><t></t><d></d></i>
+    <i n="remoting"><t></t><d></d></i>
+    ...
+</list>
+\end{lstlisting}
+
+\paragraph{Error Handling}
+When run in completion mode, the compiler will not display errors but instead try to ignore them or recover from them. 
+
+In case a not recoverable error occur while getting completion, the Haxe compiler will print the error message instead of the completion output. You can then treat any not-XML output as an error message which prevent the completion from being performed.
+
+\paragraph{Compilation cache server}
+\since{2.09}
+
+To get best speed for both compilation and completion, you can use the \ic{--wait} commandline parameter to start a Haxe compilation server. You can also use \ic{-v} to have the server printing the log. Here's an example:
+
+\begin{lstlisting}
+haxe -v --wait 6000
+\end{lstlisting}
+
+You can then connect to the Haxe server, send commandline parameters followed by a 0 byte, then read the response (either completion result or errors).
+
+You can use the \ic{--connect} commandline parameter to have haxe send its compilation commands to the server instead of executing them directly :
+
+\begin{lstlisting}
+haxe --connect 6000 myproject.hxml
+\end{lstlisting}
+
+Please notice that you can use \ic{--cwd} as the first sent commandline parameter in order to change the Haxe server current working directory, since usually class paths and other files are relative to your project.
```

Comment:
"Please note that" Review "at the first sent..." (is there a way to make this clearer?) "Haxe server's current working"